File: Week1/MathProblem/LcmAndGcd.java
public class LcmAndGcd {

/* 
* Find gcd and lcm of 2 number.
* https://www.geeksforgeeks.org/problems/lcm-and-gcd4516/1
* T(c) -> O(log(min(A,B)))
*/
    static Long[] lcmAndGcd(Long A , Long B) {

        long divisor = Math.min(A,B);
        long divident = Math.max(A,B);
        
        /* Using Euclidean algorithm or division method to find gcd. */
        while(divident % divisor != 0)
        {
            long remainder = divident % divisor;
            divident = divisor;
            divisor = remainder;
        }
        
        long gcd = divisor;
        
        /* Used property GCD(A,B)×LCM(A,B)=A×B to find the LCM (Least Common Multiple). */
        long lcm = (A*B)/gcd;
        
        Long[] arr = new Long[2];
        arr[0]=lcm;
        arr[1]=gcd;

        return arr;
    }

    public static void main(String args[])
    {
        long A = 14;
        long B = 8;
        Long[] ptr = lcmAndGcd(A,B);
                
        System.out.println("LCM = " + ptr[0] + " , GCD = "+ ptr[1]);

    }
}

File: Week1/MathProblem/Python/LcmAndGcd.py
"""
    Given two integers a and b, write a function lcmAndGcd() to compute their LCM and GCD. The function takes two integers a and 
    b as input and returns a list containing their LCM and GCD.

    https://www.geeksforgeeks.org/problems/lcm-and-gcd4516/1
"""

def lcmAndGcd(A , B):
    """ T(c) -> O(log(min(a,b))) , S(c) -> O(1) """
    divisor = min(A, B)
    divident = max(A, B)
    
    # using euclidean algorithm or division method to find GCD
    while divisor != 0:
        remainder = divident % divisor
        divident = divisor
        divisor = remainder
        
    gcd = divident
    
    # using GCD(A,B) * LCM(A,B) = A*B for calculating LCM
    lcm = (A * B) // gcd
    
    return lcm, gcd


def lcmAndGcd2(a , b):
    """ T(c) -> O(log(min(a,b))) , S(c) -> O(log(min(a,b))) """
    
    gcd = gcdRecursive(a, b)
    
    # using formula lcm * gcd = a* b
    lcm = (a*b) // gcd
    
    return [lcm, gcd]
        
def gcdRecursive(a, b):
    # using euclidean algo
    if b == 0:
        return a

    return gcdRecursive(b, a%b)


arr = lcmAndGcd(14, 8)
print(f"LCM = {arr[0]} and GCD = {arr[1]}")
arr = lcmAndGcd2(5, 0)
print(f"LCM = {arr[0]} and GCD = {arr[1]}")